### DISCUSSION TITLE
NER training is not respecting Tok2Vec in spaCy v3

### DISCUSSION THREAD

**Discussion by bratao** [explosion/spaCy#6559] [Help: Coding & Implementations]
Hello, 

Those last days I´m migrating a pipeline to spaCy v3, and I been super excited by the performance. I´m doing sequence labeling on super large texts.

But I been stuck for a couple of days on a problem, that looks like a bug ( or I´m doing something wrong).

One important requirement of my problem it that the token formatting (such as is_bold, is_italic) is important for prediction. 
I modified Token2Vec to return those extra features (as https://github.com/explosion/spaCy/discussions/6527). But for simplicity I will utilize the default token features. 

For example, my problem is the follow:

text = ["a", "a", "a", "a", "a", "a", "a", "a"]
tags = ['U-MATCH', 'B-OTHER', 'I-OTHER', 'I-OTHER', 'I-OTHER', 'I-OTHER', 'I-OTHER', 'L-OTHER']
feat = ["1", "0", "0", "0", "0", "0", "0", "0"]

Even though the words are the same, I expect that the NER learn that when the token feat=1 it is a MATCH entity, otherwise it is an OTHER entity

I created a document as 
`Doc(nlp.vocab, words=text, tags=feat)`

And modified the Tok2Vec to utilize only the TAG column

```
....
build_Tok2Vec_model(
        embed=MultiHashEmbed(
            width=128,
            rows=[500],
            attrs=["TAG"],

            include_static_vectors=False,
        ),
....

```
I manually checked the tok2vec output on a debbuger and can confirm that it is creating embedding based purely on the TAG column. 

However the training will not converge. If I do change the text to:
text = ["b", "a", "a", "a", "a", "a", "a", "a"]

It will converge immediately.

Did I did something wrong?

## How to reproduce the behavior

The following code:
```
from typing import List

from spacy import Language
from spacy.ml import MultiHashEmbed, MaxoutWindowEncoder, build_Tok2Vec_model, Floats2d, Model
from spacy.training import Example, biluo_tags_to_spans
from spacy.tokens import Doc
from spacy.util import registry


@registry.architectures.register("my_tok2vec")
def build_hash_embed_cnn_tok2vec(
) -> Model[List[Doc], List[Floats2d]]:
    return build_Tok2Vec_model(
        embed=MultiHashEmbed(
            width=128,
            rows=[500],
            attrs=["TAG"],

            include_static_vectors=False,
        ),
        encode=MaxoutWindowEncoder(
            width=128, depth=2, maxout_pieces=2, window_size=1
        ),
    )


nlp = Language()

config = {"model": {"@architectures": "my_tok2vec"}}
tok = nlp.add_pipe("tok2vec", config=config)

ner = nlp.add_pipe("ner")
train_examples = []


text = ["a", "a", "a", "a", "a", "a", "a", "a"] * 2
tags = ['U-MATCH', 'B-OTHER', 'I-OTHER', 'I-OTHER', 'I-OTHER', 'I-OTHER', 'I-OTHER', 'L-OTHER'] * 2
feat = ["1", "0", "0", "0", "0", "0", "0", "0"] * 2


doc_spacy = Doc(nlp.vocab, words=text, tags=feat)
doc_reference = Doc(nlp.vocab, words=text, tags=feat)

spans = biluo_tags_to_spans(doc_reference, tags)
doc_reference.set_ents(spans)

example = Example(doc_spacy, doc_reference)
train_examples.append(example)

ner.add_label("MATCH")
ner.add_label("OTHER")

optimizer = nlp.initialize(get_examples=lambda: train_examples)
test_examples = [x.copy() for x in train_examples]


print("This will fail to converge")

for i in range(100):
    losses = {}
    nlp.update(train_examples, drop=0.5, sgd=optimizer, losses=losses)

scores = nlp.evaluate(test_examples)

print(f"{scores['ents_f']}")

print("This will converge, even that our tok2vec is just using TAG as attribute")

text = ["b", "a", "a", "a", "a", "a", "a", "a"] * 2
train_examples = []
doc_spacy = Doc(nlp.vocab, words=text, tags=feat)
doc_reference = Doc(nlp.vocab, words=text, tags=feat)

spans = biluo_tags_to_spans(doc_reference, tags)
doc_reference.set_ents(spans)

example = Example(doc_spacy, doc_reference)
train_examples.append(example)
test_examples = [x.copy() for x in train_examples]

for i in range(100):
    losses = {}
    nlp.update(train_examples, drop=0.5, sgd=optimizer, losses=losses)

scores = nlp.evaluate(test_examples)

print(f"{scores['ents_f']}")
```

It will print:
```
This will fail to converge
0.4
This will converge, even that our tok2vec is just using TAG as attribute
1.0
```

I Would expect that both would convege to the right solution. 

## Your Environment
Python 3.8 with Spacy compiled from Today (12/12/2020) v3 develop branch version.

**Comment by bratao** [comment]
Sorry everyone, I finally got it to work. 
In my mind the NER pipe would utilize the existing tok2vec pipe before it. But after debugging a little more I understand that the NER pipe create it own tok2vec:
So we need to specify in the config:

```
config_ner = {"model": {"tok2vec": {"@architectures": "my_tok2vec"}}}
nlp.add_pipe("ner", config=config_ner)
```

Then, it works great!!

**Comment by svlandeg** [comment]
Thanks for reporting back with the solution! I've converted your Issue to a Discussion thread, as we're hoping to start a little knowledgebase of questions & answers & trouble shooting on this GH discussions board (a new feature in GH!).

I also wanted to point out that you can use the new "Listener" architecture in v3. See https://nightly.spacy.io/usage/embeddings-transformers#embedding-layers. If you use a [Tok2VecListener](https://nightly.spacy.io/api/architectures#Tok2VecListener) as `tok2vec` argument of your NER model, the listener will fetch the Tok2Vec results from the upstream Tok2Vec component. This also allows you to easily share the same Tok2Vec layer (or transformer) in multiple downstream components.
